test/functional/collection_deserialization_tests.js

```diff
@@ -0,0 +1,170 @@
+'use strict';
+var test = require('./shared').assert;
+var setupDatabase = require('./shared').setupDatabase;
+
+function User(doc) {
+  this.firstName = doc.firstName;
+  this.lastName = doc.lastName;
+}
+
+User.prototype.getFullName = function() {
+  return this.firstName + ' ' + this.lastName;
+};
+
+function serializeToUser(doc) {
+  return new User(doc);
+}
+
+describe('Collection (de)serialization', function() {
+  before(function() {
+    return setupDatabase(this.configuration);
+  });
+
+  it('should deserialize find', {
+    metadata: {
+      requires: { topology: ['single', 'replicaset', 'sharded', 'ssl', 'heap', 'wiredtiger'] }
+    },
+
+    // The actual test we wish to run
+    test: function(done) {
+      var configuration = this.configuration;
+      var ObjectID = configuration.require.ObjectID;
+
+      var client = configuration.newClient(configuration.writeConcernMax(), {
+        poolSize: 1
+      });
+
+      client.connect(function(err, client) {
+        var db = client.db(configuration.db);
```

Comment:
This line should be switched with the next one as follows:

```js
client.connect(function(err, client) {
  expect(err).to.eql(null);
  const db = client.db(configuration.db);
  ...
```

If there is an error, `client` will be `null|undefined`, and the test will error out with the message `TypeError: Cannot read property 'db' of undefined`. By switching the lines, the error message will be clearer.